From 1b66d824326abf5be42e5f33979059fb8f1fd6a8 Mon Sep 17 00:00:00 2001
From: Marc Mao <xmao4@android.intel.com>
Date: Tue, 1 Jun 2021 19:59:34 +0800
Subject: [PATCH] Add external buffer tiling flag for tiled format

TODO: to support tiling format better, we should use modifier
information with PRIME2 API.

Change-Id: I39a79b37bfa5658230ebb42846dc0651fba31aa4
Signed-off-by: Marc Mao <xmao4@android.intel.com>
---
 omx_utils/src/mfx_omx_vaapi_allocator.cpp | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/omx_utils/src/mfx_omx_vaapi_allocator.cpp b/omx_utils/src/mfx_omx_vaapi_allocator.cpp
index f942755..43c9417 100644
--- a/omx_utils/src/mfx_omx_vaapi_allocator.cpp
+++ b/omx_utils/src/mfx_omx_vaapi_allocator.cpp
@@ -103,6 +103,19 @@ unsigned int ConvertGrallocFourccToVAFormat(int fourcc)
     }
 }
 
+// quick fix tiling issue, formal should use modifier with PRIME2 interface
+bool isTiledGrallocFourcc(int fourcc) {
+    switch (fourcc)   {
+        case HAL_PIXEL_FORMAT_NV12_Y_TILED_INTEL:
+        case HAL_PIXEL_FORMAT_YUV420PackedSemiPlanar_Tiled_INTEL:
+        case HAL_PIXEL_FORMAT_P010_INTEL:
+            return true;
+        default:
+            break;
+    }
+    return false;
+}
+
 // We need to sort surface IDs before destroy.
 bool compareMid(const vaapiMemId* first, const vaapiMemId* second)
 {
@@ -530,6 +543,8 @@ mfxStatus MfxOmxVaapiFrameAllocator::CreateSurfaceFromGralloc(const mfxU8* handl
 #ifdef MFX_OMX_USE_PRIME
     surfExtBuf.buffers = (uintptr_t *)&(info.prime);
     surfExtBuf.flags = VA_SURFACE_ATTRIB_MEM_TYPE_DRM_PRIME;
+    if (isTiledGrallocFourcc(info.format))
+        surfExtBuf.flags  |= VA_SURFACE_EXTBUF_DESC_ENABLE_TILING;
 #else
     surfExtBuf.buffers = (uintptr_t *)&handle;
     surfExtBuf.flags = VA_SURFACE_ATTRIB_MEM_TYPE_ANDROID_GRALLOC;
-- 
2.35.1

